### ISSUE TITLE
Is it possible to download individual archives, list them? 

### ISSUE THREAD

**Issue by shekarkcb** [vsespb/mt-aws-glacier#3]
Hi,

This Perl implementation module is working great - (This is not a issue list, but an feature requests :) )

a). I am able to upload the files inside the directory.
b). Create download request for the files (gave max-file=1 )
c). After 4hrs was able to download the file.

but just curious, 

a). Is it possible to download individual archives?  by name or some other way? 
b). Delete individual files ( by file name or initiating the request for download )
c). List the archives per vault
d). Create/Delete vault
e). Handle error message properly, i.e if i run script without any options/short of options it just dies.

Just a request, but not issues list. 

Thanks.

**Comment by vsespb** [comment]
Thanks for your feedback.

a) not yet, but will be implemented for sure
b) not yet, but will be implemented for sure
c) not yet. I am planing to implement saving filenames in vault and a command to restore Journal from vault's listing (be careful, as mentioned in README this program is Beta yet also you'll need to save your Journal in a safe place to restore your backup)
d) probably yes, but it's low priority I think.
e) will be implemented

priorities are:  c) e) a) b) d)

I think (c) will be implemented in a week or so.

**Comment by major12** [comment]
Could you create some low-level actions? Or even a separate script with them.
It is good to have high-level action like "sync", but it will be very useful to have some basic command like
upload (single-file, vault as parameter)
get-list
prepare-download
download

For some cases creating directory, moving/copying backup file here, saving a journal after upload, etc are unnecessary overhead. 

I saw Net::Amazon::Glacier at CPAN, but it cannot handle files larger then 4Gb and it has a lot of dependencies.

**Comment by vsespb** [comment]
Yes, panned too. I started from high level function, because there are other linux cli software which supports low level functions.

**Comment by bembol** [comment]
Just a quick question. When we can expect additional functionality like:
1. Delete individual files based on archive id of file name if possible
2. Download individual files based on archive id of file name if possible

**Comment by vsespb** [comment]
I think 1-1.5 months

**Comment by dc65** [comment]
What exacly does the "restore" command do? Does it restore all files listed in the specified journal, or does it first compare with the files in --from-dir to see which files are missing? My purpose is to implement some workaround to download one specified archive by filename, while I am waiting for your feature to be implemented. I would experiment, but with Glacier restore fees, it is far cheaper to ask first :-) Thanks for your software, very much appreciated.

**Comment by vsespb** [comment]
It does first compare with the files in --from-dir to see which files are missing, then it restore files (to download files wait 4h and use restore-completed).

To download one archive you can use the workaround: using `grep` you can extract from Journal only record (one line is one record) corresponding to that archive to another new temporary journal file. And use that new file for restoring.
You can try grep filename or grep archive id.

**Comment by vsespb** [comment]
Update:

Upload single file or file from STDIN implemented (but only together with Journal functionality). Multithreaded. Without intermediate file buffering.
Download of single files not implemented yet.

see `upload-file` command in README

**Comment by 108adams** [comment]
Hello. What's the status of deleting an archive functionality?

**Comment by vsespb** [comment]
Hello.
1. You can delete individual archive(s) now using `purge-vault` command  with `--filter` options. (only filter by filename, only if you already have Journal for files).
2. If you need to delete file by archive id, you can workaround (you still need journal):
   
   ```
   grep MY_ARCHIVE_ID journal > partial_journal
   ```
   
   and then
   `purge-vault` with partial_journal.
3. If your files originally were uploaded by 3rd party tools, filenames in journal are equal to archive id, so you can try use (1) with them.

But a dedicated command for deleting individual  archives (with or without journal) is not implemented yet.
